run/deadline.go

```diff
@@ -0,0 +1,25 @@
+package run
+
+import (
+	"context"
+	"time"
+)
+
+// WithDeadline runs the given function passing it a context with the deadline adjusted
+// to be no later than the provided deadline.  The calling function should respect
+// the context.
+func WithDeadline(deadline time.Time, fn func(context.Context) error) func(context.Context) error {
+	return func(ctx context.Context) error {
+		ctx, cancel := context.WithDeadline(ctx, deadline)
```

Comment:
why not doing simply this?
```
		ctx, cancel := context.WithDeadline(ctx, deadline)
		defer cancel()
		return fn(ctx)
```

Reply:
Because the context will be cancelled when exiting the function and the function won't run

Reply:
but also with the original code the function will end before exiting. If you replace it with this code, it passes the test.

Reply:
Ok, I see. Let's consider this other alternative them, which is more simpler  as you commented. Thanks